diff --git a/MdeModulePkg/Bus/Pci/PciBusDxe/PciResourceSupport.c b/MdeModulePkg/Bus/Pci/PciBusDxe/PciResourceSupport.c
index 8ffd05f..30ed2c6 100644
--- a/MdeModulePkg/Bus/Pci/PciBusDxe/PciResourceSupport.c
+++ b/MdeModulePkg/Bus/Pci/PciBusDxe/PciResourceSupport.c
@@ -1276,6 +1276,14 @@ ProgramBar (
 
   Address = Base + Node->Offset;
 
+  //
+  // Preserve preset BAR address for VFIO passthrough devices.
+  // QEMU presets the host BAR address in config space for such devices.
+  //
+  if (Node->PciDev->PciBar[Node->Bar].BaseAddress != 0) {
+    Address = Node->PciDev->PciBar[Node->Bar].BaseAddress;
+  }
+
   //
   // Indicate pci bus driver has allocated
   // resource for this device
@@ -1365,6 +1373,14 @@ ProgramVfBar (
 
   Address = Base + Node->Offset;
 
+  //
+  // Preserve preset VF BAR address for VFIO passthrough devices.
+  // QEMU presets the host BAR address in config space for such devices.
+  //
+  if (Node->PciDev->VfPciBar[Node->Bar].BaseAddress != 0) {
+    Address = Node->PciDev->VfPciBar[Node->Bar].BaseAddress;
+  }
+
   //
   // Indicate pci bus driver has allocated
   // resource for this device
